### PR TITLE
[SPARK-35628][SS] RocksDBFileManager - load checkpoint from DFS

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
The implementation for the load operation of RocksDBFileManager.

### Why are the changes needed?
Provide the functionality of loading all necessary files for specific checkpoint versions from DFS to the given local directory.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New UT added.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 1]
```diff
@@ -20,12 +20,125 @@ package org.apache.spark.sql.execution.streaming.state
...
+      generateFiles(verificationDir, Seq(
+        "sst-file1.sst" -> 11, // files with same exist but different sizes, should get overwritten
+        "other-file1" -> 101,
+        "archive/00001.log" -> 1001,
+        "random-sst-file.sst" -> 100, // unnecessary files, should get deleted
+        "random-other-file" -> 9,
+        "00005.log" -> 101,
+        "archive/00007.log" -> 101
+      ))
+      loadAndVerifyCheckpointFiles(fileManager, verificationDir, version = 1, cpFiles1, 101)
```
Comment (xuanyuanking) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:79]: Per https://github.com/apache/spark/pull/32582#discussion_r642036286

**Comment by xuanyuanking** [comment]
Rebased this PR based on #32582. It's ready for review now. cc @viirya and @HeartSaVioR

**Comment by viirya** [comment]
Thank you @xuanyuanking. I'll find some time to review this.

**Comment by xuanyuanking** [comment]
@viirya Great thanks for your help!

**Comment by xuanyuanking** [comment]
To make the RocksDB state store implementation can be reviewed quickly and easily. I just created all the rest PRs to provide us a global perspective. We can review them one by one, and I'll keep updating each of them:
https://github.com/apache/spark/pull/32928 - [SPARK-35784][SS] Implementation for RocksDB instance
https://github.com/apache/spark/pull/32933 - [SPARK-35785][SS] Cleanup support for RocksDB instance
https://github.com/apache/spark/pull/32934 - [SPARK-35788][SS] Metrics support for RocksDB instance

cc @viirya and @HeartSaVioR Thanks for your review.

**Comment by viirya** [comment]
Thanks you @xuanyuanking!

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -3084,6 +3085,35 @@ private[spark] object Utils extends Logging {
     conf.get(config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT)
       .getOrElse(Utils.timeStringAsMs(s"${conf.get(Network.NETWORK_TIMEOUT)}s"))
   }
+
+  /** Decompress a zip file into a local dir. File names are read from the zip file. */
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: Looks like this is not the Java doc style we follow.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -3084,6 +3085,35 @@ private[spark] object Utils extends Logging {
...
+  def unzipFromFile(fs: FileSystem, dfsZipFile: Path, localDir: File): Seq[File] = {
+    val files = new mutable.ArrayBuffer[File]()
+    val in = new ZipInputStream(fs.open(dfsZipFile))
+    var out: OutputStream = null
+    try {
+      var entry = in.getNextEntry()
+      while (entry != null) {
+        if (!entry.isDirectory) {
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3097]: We don't process directory. Could you also mention it in the method doc?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 4]
```diff
@@ -3084,6 +3085,35 @@ private[spark] object Utils extends Logging {
...
+        }
+        entry = in.getNextEntry()
+      }
+      in.close() // so that any error in closing does not get ignored
+      logInfo(s"Unzipped from $dfsZipFile\n\t${files.mkString("\n\t")}")
+    } finally {
+      // Close everything no matter what happened
+      IOUtils.closeQuietly(in)
+      IOUtils.closeQuietly(out)
+    }
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3114]: Hmm, are we sure we don't need to process any error during unzipping?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 5]
```diff
@@ -200,6 +246,55 @@ class RocksDBFileManager(
...
+    val requiredFileNameToFileDetails = immutableFiles.map(f => f.localFileName -> f).toMap
+    // Delete unnecessary local immutable files
+    listRocksDBFiles(localDir)._1
+      .foreach { existingFile =>
+        val isSameFile =
+          requiredFileNameToFileDetails.get(existingFile.getName).exists(_.isSameFile(existingFile))
+        if (!isSameFile) {
+          existingFile.delete()
+          logInfo(s"Deleted local file $existingFile")
+        }
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:265]: When it is possible to have existing file in local dir which has same file name but not the same file in DFS?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:265]: Or just a safer guard?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 6]
```diff
@@ -200,6 +246,55 @@ class RocksDBFileManager(
...
+        val dfsFile = dfsFilePath(file.dfsFileName)
+        // Note: The implementation of copyToLocalFile() closes the output stream when there is
+        // any exception while copying. So this may generate partial files on DFS. But that is
+        // okay because until the main [version].zip file is written, those partial files are
+        // not going to be used at all. Eventually these files should get cleared.
+        fs.copyToLocalFile(dfsFile, new Path(localFile.getAbsoluteFile.toURI))
+        val localFileSize = localFile.length()
+        filesCopied += 1
+        bytesCopied += localFileSize
+        logInfo(s"Copied $dfsFile to $localFile - $localFileSize bytes")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:284]: We can do this logInfo after the file size check.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 7]
```diff
@@ -200,6 +246,55 @@ class RocksDBFileManager(
...
+        if (localFileSize != expectedSize) {
+          throw new IllegalStateException(
+            s"Copied $dfsFile to $localFile," +
+              s" expected $expectedSize bytes, found $localFileSize bytes ")
+        }
+      } else {
+        filesReused += 1
+      }
+    }
+    logInfo(s"Copied $filesCopied files ($bytesCopied bytes) from DFS to local.")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:295]: Add `filesReused` into this log message?

**Review by viirya — COMMENTED** [review]
Overall looks good. Just a few questions and suggestions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 8]
```diff
@@ -153,6 +156,49 @@ class RocksDBFileManager(
...
+      versionToRocksDBFiles.put(version, metadata.immutableFiles)
+      metadataFile.delete()
+      metadata
+    }
+    logFilesInDir(localDir, s"Loaded checkpoint files for version $version")
+    metadata
+  }
+
+  /** Get the latest version available in the DFS directory. If no data present, it returns 0. */
+  def getLatestVersion(): Long = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:190]: Is this only used by tests?

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: We have been using one-liner java doc; e.g. https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala#L50

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3097]: It'd be ideal if we make it clear in method name, like `unzipFilesFromFile`.  (Ideally I'd like to see this also extracts the directory, but let's postpone it till necessary.)

In general we expect unzipping will extract the directories as well. That said, we need to make the behavior very clear to the caller side. I agree this should be mentioned to the java doc, but method name should be also intuitive to expect the actual behavior.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 4 above)
Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3114]: This looks safe; if there's an exception we may see some files being extracted and the one of output files may be broken, but callers will catch an exception and indicate the output directory is not healthy. If necessary let's document this in javadoc as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 9]
```diff
@@ -153,6 +156,49 @@ class RocksDBFileManager(
     logInfo(s"Saved checkpoint file for version $version")
   }
 
+  /**
+   * Load all necessary files for specific checkpoint version from DFS to given local directory.
+   * If version is 0, then it will deleted all files in the directory. For other versions, it
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:161]: nit: deleted -> delete

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 10]
```diff
@@ -153,6 +156,49 @@ class RocksDBFileManager(
...
+   * local directory.
+   */
+  def loadCheckpointFromDfs(version: Long, localDir: File): RocksDBCheckpointMetadata = {
+    logInfo(s"Loading checkpoint files for version $version")
+    val metadata = if (version == 0) {
+      if (localDir.exists) Utils.deleteRecursively(localDir)
+      localDir.mkdirs()
+      RocksDBCheckpointMetadata(Seq.empty, 0)
+    } else {
+      // Delete all non-immutable files in local dir, and unzip new ones from DFS commit file
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:172]: What if we just remove the all files in localDir? Just would like to know the reason we don't clear the directory but just remove the specific files. Would we need to leverage some remaining files?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 6 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:284]: More specifically, we can do the file size check just after copyToLocalFile, and accumulations can be placed later.

**Review by HeartSaVioR — COMMENTED** [review]
Minor comments.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: Oh, I thought it is two lines.

Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: But we'd like to see this multiple lines as below review comments :)

Comment (viirya) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: yea :)

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3089]: Yea, we'll have multiple lines here in the next commit :)

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3097]: Make sense, method name changed and comment added.

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 4 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3114]: Yes, we rely on the caller side to address any exceptions. Javadoc added as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 5 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:265]: A safer guard for checking both file names and file size.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 6 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:284]: Thanks, do the size check right after copyToLocalFile and place the logInfo in the end.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 7 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:295]: Thanks! Done in the next commit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 8 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:190]: Yes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 9 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:161]: Thanks, done in the next commit

**Comment by xuanyuanking** [comment]
Great thanks for your detailed review, @viirya @HeartSaVioR. All comments addressed.

**Review by HeartSaVioR — APPROVED** [review]
+1
Let's wait for @viirya to have another round of review and do explicit approval.

**Comment by HeartSaVioR** [comment]
@xuanyuanking Oh we need to fix new code conflicts as well.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala [hunk 11]
```diff
@@ -53,10 +62,24 @@ class RocksDBSuite extends SparkFunSuite {
...
       )
       saveCheckpointFiles(fileManager, cpFiles1, version = 1, numKeys = 101)
+      assert(fileManager.getLatestVersion() === 1)
       assert(numRemoteSSTFiles == 2) // 2 sst files copied
       assert(numRemoteLogFiles == 2) // 2 log files copied
 
-      // Save SAME version again with different checkpoint files and verify
+      // Load back the checkpoint files into another local dir with existing files and verify
+      generateFiles(verificationDir, Seq(
+        "sst-file1.sst" -> 11, // files with same exist but different sizes, should get overwritten
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:71]: files with same? with same filename?

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
@xuanyuanking Could you resolve the conflict and the minor comment? Then we can move this forward. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala (hunk 11 above)
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala]: Ah thanks! Fix done.

**Comment by xuanyuanking** [comment]
Thanks for the review!

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3126]: @xuanyuanking Seems you overwrite previous change? This looks the previous version.

Comment (HeartSaVioR) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3126]: @xuanyuanking friendly reminder

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3126]: ……Sorry I missed this... Updating

**Comment by HeartSaVioR** [comment]
Jenkins passed. Thanks! Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks @xuanyuanking for the contribution! I merged into master.

**Comment by HeartSaVioR** [comment]
Please also rebase the next PRs into master branch; we can continue reviewing next PR.

**Comment by xuanyuanking** [comment]
Copy that. I'm rebasing now. Thanks for the help!

**Comment by HeartSaVioR** [comment]
UPDATE: we found a consistent break on Scala 2.13 build caused by this. @xuanyuanking is working on the fix so please allow us some time to fix it as follow-up PR instead of reverting this.

**Comment by xuanyuanking** [comment]
Great thanks @HeartSaVioR and @Ngone51. Submitted https://github.com/apache/spark/pull/33084.

**Pull request closed without merging**